openfisca_core/data_storage.py

```diff
@@ -1,136 +1,191 @@
-# -*- coding: utf-8 -*-
-
-import shutil
+import abc
 import os
+import shutil
+from typing import Dict, List, Optional
 
-import numpy as np
+import numpy
 
+from openfisca_core import indexed_enums
 from openfisca_core import periods
-from openfisca_core.periods import ETERNITY
-from openfisca_core.indexed_enums import EnumArray
 
+from typing_extensions import Protocol
+
+
+class StorageLike(Protocol):
+
+    @abc.abstractmethod
+    def get(self, period: periods.Period) -> numpy.ndarray:
+        ...
+
+    @abc.abstractmethod
+    def put(self, value: numpy.ndarray, period: periods.Period) -> None:
+        ...
 
-class InMemoryStorage(object):
+    @abc.abstractmethod
+    def delete(self, period: Optional[periods.Period] = None) -> None:
+        ...
+
+    @abc.abstractmethod
+    def get_known_periods(self) -> List[periods.Period]:
+        ...
+
+    @abc.abstractmethod
+    def get_memory_usage(self) -> Dict[str, int]:
+        ...
+
+    def _pop(self, period: periods.Period, items: list) -> dict:
+        return {item: value for item, value in items if not period.contains(item)}
+
+
+class InMemoryStorage(StorageLike):
     """
-    Low-level class responsible for storing and retrieving calculated vectors in memory
+    Low-level class responsible for storing and retrieving calculated vectors in memory.
     """
 
-    def __init__(self, is_eternal = False):
+    _arrays: dict
+    is_eternal: bool
+
+    def __init__(self, is_eternal: bool = False) -> None:
         self._arrays = {}
         self.is_eternal = is_eternal
 
-    def get(self, period):
+    def get(self, period: periods.Period) -> numpy.ndarray:
         if self.is_eternal:
-            period = periods.period(ETERNITY)
-        period = periods.period(period)
+            period = periods.period(periods.ETERNITY)
 
+        period = periods.period(period)
         values = self._arrays.get(period)
+
         if values is None:
             return None
+
         return values
 
     def put(self, value, period):
         if self.is_eternal:
-            period = periods.period(ETERNITY)
-        period = periods.period(period)
+            period = periods.period(periods.ETERNITY)
 
+        period = periods.period(period)
         self._arrays[period] = value
 
-    def delete(self, period = None):
+    def delete(self, period: Optional[periods.Period] = None) -> None:
         if period is None:
             self._arrays = {}
             return
 
         if self.is_eternal:
-            period = periods.period(ETERNITY)
+            period = periods.period(periods.ETERNITY)
+
         period = periods.period(period)
 
-        self._arrays = {
-            period_item: value
-            for period_item, value in self._arrays.items()
-            if not period.contains(period_item)
-            }
+        if period is not None:
```

Comment:
This null check is redundant given the constructor call above. (Which itself is redundant given the guard clause.)